File: JAVA/sortedArrayToBST.java
package edu.ucsc;

/**
 * Leetcode Number:
 * Problem Name:
 * @author: Created by Xiaotong Li 
 * @time: 5/28/18:2018-05-28 17:24.
 */

public class sortedArrayToBST {
    public TreeNode sortedArrayToBST(int[] num){
        if (num.length == 0){
            return null;
        }
        return helper(num, 0, num.length - 1);
    }

    private TreeNode helper(int[] num, int low, int high ){
        if (low > high - 1){
            return null;
        }
        int mid = (low + high)/2;
        TreeNode node = new TreeNode(num[mid]);
        node.left = helper(num, low, mid - 1);
        node.right = helper(num, mid + 1, high);
        return node;
    }
}

File: Python/BinarySearchTree/sortedArrayToBST.py
#! /Users/xiaotongli/anaconda3/bin/python
# -*- coding: utf-8 -*-
# @Time    : 10/12/18 10:55 PM
# @Author  : Xiaotong Li
# @School  : University of California, Santa Cruz
# @FileName: sortedArrayToBST.py
# @Software: PyCharm


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if not nums:
            return None

        mid = len(nums) // 2

        root = TreeNode(nums[mid])
        root.left = self.sortedArrayToBST(nums[:mid])
        root.right = self.sortedArrayToBST(nums[mid + 1:])

        return root
